### ISSUE TITLE
[Flaky CI] Can't connect to websocket (500): websocket: bad handshake

### ISSUE THREAD

**Issue by jean-edouard** [kubevirt/kubevirt#3873]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
/triage build-officer

**What happened**:
Flaky test:
```
[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:component]Networking Multiple virtual machines connectivity using bridge binding interface VirtualMachineInstance with default interface model [test_id:1550]should expose the right device type to the guest
```
is this prow job:
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/3856/pull-kubevirt-e2e-k8s-1.18/1285864699644088320

**What you expected to happen**:
Test to pass

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
- Kubernetes version (use `kubectl version`):
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by stu-gott** [comment]
Issue observed again on 2020-07-24:

```
 • Failure [56.383 seconds]
[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:component]VirtualMachine
/root/go/src/kubevirt.io/kubevirt/tests/vm_test.go:53
  A valid VirtualMachine given
  /root/go/src/kubevirt.io/kubevirt/tests/vm_test.go:266
    Using virtctl interface
    /root/go/src/kubevirt.io/kubevirt/tests/vm_test.go:748
      Using RunStrategyAlways
      /root/go/src/kubevirt.io/kubevirt/tests/vm_test.go:853
        [test_id:3165]should restart a succeeded VMI [It]
        /root/go/src/kubevirt.io/kubevirt/tests/vm_test.go:932
        Unexpected error:
      
            <*kubecli.AsyncSubresourceError | 0xc0028b3360>: {
                err: "Can't connect to websocket (500): websocket: bad handshake\n",
                StatusCode: 500,
            }
            Can't connect to websocket (500): websocket: bad handshake
            
        occurred
        /root/go/src/kubevirt.io/kubevirt/tests/vm_test.go:946 
```

Prow job: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-kubevirt-hyperconverged-cluster-operator-release-4.5-hco-e2e-nightly-bundle-release-4-5-azure4/1286496876157210624#1:build-log.txt%3A646

**Comment by nunnatsa** [comment]
And again on 02-Aug-2020:
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-kubevirt-hyperconverged-cluster-operator-release-4.5-hco-e2e-nightly-bundle-release-4-5-azure4/1289758438993170432

**Comment by oshoval** [comment]
Again at 7.9.2020
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4099/pull-kubevirt-check-tests-for-flakes/1302849844318048258

**Comment by fgimenez** [comment]
Happened here again https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4398/pull-kubevirt-e2e-k8s-1.18/1319596819776278528

In the kube-apiserver pod logs there are these errors at the exact same time of the failure, `14:39:51`:
```
W1023 14:39:51.356968       1 dispatcher.go:169] Failed calling webhook, failing open virtualmachineinstances-mutator.kubevirt.io: failed calling webhook "virtualmachineinstances-mutator.kubevirt.io": Post https://virt-api.kubevirt.svc:443/virtualmachineinstances-mutate?timeout=30s: x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "kubevirt.io")
E1023 14:39:51.357930       1 dispatcher.go:170] failed calling webhook "virtualmachineinstances-mutator.kubevirt.io": Post https://virt-api.kubevirt.svc:443/virtualmachineinstances-mutate?timeout=30s: x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "kubevirt.io")
``` 
virt-api certificate is stored in the kubevirt-virt-api-certs secret, looking at the secrets log https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4398/pull-kubevirt-e2e-k8s-1.18/1319596819776278528/artifacts/k8s-reporter/1_secrets.log
```
{
            "metadata": {
                "name": "kubevirt-virt-api-certs",
                "namespace": "kubevirt",
                "selfLink": "/api/v1/namespaces/kubevirt/secrets/kubevirt-virt-api-certs",
                "uid": "006526fd-6cbe-4105-ae21-5ddead62d28e",
                "resourceVersion": "93063",
                "creationTimestamp": "2020-10-23T11:41:40Z",
                "labels": {
                    "app.kubernetes.io/component": "kubevirt",
                    "app.kubernetes.io/managed-by": "kubevirt-operator"
                },
                "annotations": {
                    "kubevirt.io/customizer-identifier": "bf21a9e8fbc5a3846fb05b4fa0859e0917b2202f",
                    "kubevirt.io/duration": "\u0026Duration{Duration:7m0s,}",
                    "kubevirt.io/generation": "135",
                    "kubevirt.io/install-strategy-identifier": "5254e1aeb5193487c3cfdceec6af1bbc04cefaba",
                    "kubevirt.io/install-strategy-registry": "registry:5000/kubevirt",
                    "kubevirt.io/install-strategy-version": "devel"
                },
                "managedFields": [
                    {
                        "manager": "virt-operator",
                        "operation": "Update",
                        "apiVersion": "v1",
                        "time": "2020-10-23T14:39:50Z",
                        "fieldsType": "FieldsV1",
                        "fieldsV1": {
                            "f:data": {
                                ".": {},
                                "f:tls.crt": {},
                                "f:tls.key": {}
                            },
.............
```
looks like there was an update operation 1 second before the error was reported, at `14:39:50`. Might the certificate rotation be causing this problem?

Also, an additional question from my side, i see that the duration for this certificate is set to 7min for functests from the kubevirt CR instead of the default 24h, is this done intentionally to exercise certificate rotation during tests?

**Comment by akrejcir** [comment]
Happened again on 26.10.2020
https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4307/pull-kubevirt-e2e-k8s-1.17/1320617418665496576

**Comment by rmohr** [comment]
@fgimenez very good observations.

> looks like there was an update operation 1 second before the error was reported, at `14:39:50`. Might the certificate rotation be causing this problem?

That can indeed be a problem. What I find interesting is the the k8s apiserver had issues connecting. Since I think that we first update the webhook configuration with the new CA before we roll out the new client certificates. 

> 
> Also, an additional question from my side, i see that the duration for this certificate is set to 7min for functests from the kubevirt CR instead of the default 24h, is this done intentionally to exercise certificate rotation during tests?

Yes that is exactly the reason.

**Comment by oshoval** [comment]
/cc @EdDev 
saw it also on SRIOV locally and there is great info here

**Comment by rmohr** [comment]
/close

"Fixed" in #4522. It takes sometimes unusually long until the kubelet updates the certificate secrets.
